base_layer/core/src/base_node/comms_interface/inbound_handlers.rs

```diff
@@ -452,6 +446,30 @@ where B: BlockchainBackend + 'static
             );
             return Ok(());
         }
+        {
+            // we use a double lock to make sure we can only reconcile one block at a time
+            let read_lock = self.list_of_reconciling_blocks.read().await;
+            if read_lock.contains(&block_hash) {
+                debug!(
+                    target: LOG_TARGET,
+                    "Block with hash `{}` is already being reconciled",
+                    block_hash.to_hex()
+                );
+                return Ok(());
+            }
+        }
+        {
+            let mut write_lock = self.list_of_reconciling_blocks.write().await;
+            if self.blockchain_db.block_exists(block_hash).await? {
+                debug!(
+                    target: LOG_TARGET,
+                    "Block with hash `{}` already stored",
+                    block_hash.to_hex()
+                );
+                return Ok(());
+            }
+            write_lock.insert(block_hash);
+        }
```

Comment:
You can't split these. 

thread 1 wakes, read doesnt contain block X, sleeps
thread 2 wakes, read doesnt contain block X, writes hash
thread 1 wakes, writes hash and continues

These can happen in any order but only one order is valid.

Reply:
You can also use `write_lock.insert(block_hash);` to determine if the entry already exists, it returns true if the entry exists